README.md
# Specs

> **S**pecs **P**arallel **ECS**

[![Build Status][bi]][bl] [![Crates.io][ci]][cl] [![Gitter][gi]][gl] ![MIT/Apache][li] [![Docs.rs][di]][dl] [![Code coverage][coi]][cov] ![LoC][lo]

[bi]: https://travis-ci.org/amethyst/specs.svg?branch=master
[bl]: https://travis-ci.org/amethyst/specs

[ci]: https://img.shields.io/crates/v/specs.svg
[cl]: https://crates.io/crates/specs/

[li]: https://img.shields.io/crates/l/specs.svg?maxAge=2592000

[di]: https://docs.rs/specs/badge.svg
[dl]: https://docs.rs/specs/

[gi]: https://badges.gitter.im/slide-rs/specs.svg
[gl]: https://gitter.im/slide-rs/specs

[coi]: https://img.shields.io/codecov/c/gitlab/torkleyy/specs/master.svg
[cov]: https://codecov.io/gl/torkleyy/specs/branch/master

[lo]: https://tokei.rs/b1/github/slide-rs/specs?category=code

Specs is an Entity-Component System written in Rust.
Unlike most other ECS libraries out there, it provides

* easy parallelism
* high flexibility
    * contains 5 different storages for components, which can be extended by the user
    * its types are mostly not coupled, so you can easily write some part yourself and
      still use Specs
    * `System`s may read from and write to components and resources, can depend on each
      other and you can use barriers to force several stages in system execution
* high performance for real-world applications

Minimum Rust version: 1.70

## [Link to the book][book]

[book]: https://specs.amethyst.rs/docs/tutorials/

## Example

```rust
use specs::prelude::*;

// A component contains data
// which is associated with an entity.
#[derive(Debug)]
struct Vel(f32);

impl Component for Vel {
    type Storage = VecStorage<Self>;
}

#[derive(Debug)]
struct Pos(f32);

impl Component for Pos {
    type Storage = VecStorage<Self>;
}

struct SysA;

impl<'a> System<'a> for SysA {
    // These are the resources required for execution.
    // You can also define a struct and `#[derive(SystemData)]`,
    // see the `full` example.
    type SystemData = (WriteStorage<'a, Pos>, ReadStorage<'a, Vel>);

    fn run(&mut self, (mut pos, vel): Self::SystemData) {
        // The `.join()` combines multiple component storages,
        // so we get access to all entities which have
        // both a position and a velocity.
        for (pos, vel) in (&mut pos, &vel).join() {
            pos.0 += vel.0;
        }
    }
}

fn main() {
    // The `World` is our
    // container for components
    // and other resources.
    let mut world = World::new();
    world.register::<Pos>();
    world.register::<Vel>();

    // An entity may or may not contain some component.

    world.create_entity().with(Vel(2.0)).with(Pos(0.0)).build();
    world.create_entity().with(Vel(4.0)).with(Pos(1.6)).build();
    world.create_entity().with(Vel(1.5)).with(Pos(5.4)).build();

    // This entity does not have `Vel`, so it won't be dispatched.
    world.create_entity().with(Pos(2.0)).build();

    // This builds a dispatcher.
    // The third parameter of `with` specifies
    // logical dependencies on other systems.
    // Since we only have one, we don't depend on anything.
    // See the `full` example for dependencies.
    let mut dispatcher = DispatcherBuilder::new().with(SysA, "sys_a", &[]).build();
    // This will call the `setup` function of every system.
    // In this example this has no effect since we already registered our components.
    dispatcher.setup(&mut world);

    // This dispatches all the systems in parallel (but blocking).
    dispatcher.dispatch(&mut world);
}
```

Please look into [the examples directory](examples) for more.

## Public dependencies

| crate    | version                                                                                        |
|----------|------------------------------------------------------------------------------------------------|
| hibitset | [![hibitset](https://img.shields.io/crates/v/hibitset.svg)](https://crates.rs/crates/hibitset) |
| rayon    | [![rayon](https://img.shields.io/crates/v/rayon.svg)](https://crates.rs/crates/rayon)          |
| shred    | [![shred](https://img.shields.io/crates/v/shred.svg)](https://crates.rs/crates/shred)          |
| shrev    | [![shrev](https://img.shields.io/crates/v/shrev.svg)](https://crates.rs/crates/shrev)          |

## Contribution

Contribution is very welcome! If you didn't contribute before,
just filter for issues with "easy" or "good first issue" label.
Please note that your contributions are assumed to be dual-licensed under Apache-2.0/MIT.


CHANGELOG.md
# Unreleased

# 0.20.0 (2023-09-24)

* MSRV to 1.70.0 ([#765])
* Significant refactor of `Join` and related traits to alleviate soundness
  issues. Includes introduction of a lending/streaming join via the `LendJoin`
  trait which is the new common denominator implemented by joinable types.
  ([#765])

[#765]: https://github.com/amethyst/specs/pull/765

# 0.19.0 (2023-06-10)

* Bump MSRV to 1.65.0 ([#766])
* Added index where entity deletion stopped to the error returned from `WorldExt::delete_entities` ([#766])
* Fix bug where deleting an entity with the wrong generation could clear the components of an existing entity. ([#766])
* Bump shred to version `0.14.1`, MSRV to 1.60.0 ([shred changelog][shred_changelog], [#756])

[shred_changelog]: https://github.com/amethyst/shred/blob/master/CHANGELOG.md#0141-2022-07-14
[#756]: https://github.com/amethyst/specs/pull/756
[#766]: https://github.com/amethyst/specs/pull/766

# 0.18.0 (2022-07-02)

* Increase minumum rustc version to 1.56.1 because of shred and hashbrown dependencies ([#748])

[#748]: https://github.com/amethyst/specs/pull/748

Full changelog: https://github.com/amethyst/specs/compare/v0.17.0...v0.18.0

# 0.17.0 (2021-06-08)

* Deprecate `error::NoError` in favor of `std::convert::Infallible` ([#688])
* Use `#[non_exhaustive]` for `error::Error`. Note this bumps the minimum supported rust version to 1.40 ([#688]).
* Add the `derive` feature that enables all derive-related smaller features
  (`specs-derive` and `shred-derive` currently). ([#687])

[#687]: https://github.com/amethyst/specs/pull/687
[#688]: https://github.com/amethyst/specs/pull/688

# 0.16.1 (2020-02-18)

* `JoinIter` now implements `Clone` when inner types are `Clone` -- usually for immutable `join()`s. ([#620])
* Bump `hibitset` to `0.6.3`. ([#620])
* `StorageEntry::replace` replaces a component, returning the previous value if any. ([#622])

[#620]: https://github.com/slide-rs/specs/pull/620
[#622]: https://github.com/slide-rs/specs/pull/622

# 0.16.0 (2020-02-13)

* Update `syn`, `quote` and `proc-macro2` to `1.0`. ([#648])
* Implement `ParJoin` for `MaybeJoin` if the inner type is `ParJoin`. ([#655])
* Remove `"nightly"` feature -- improved panic messages are available on stable. ([#671])
* Bump `shred` to `0.10.2`. ([#671], [#674], [#683])
* Components and resources no longer need to be `Send + Sync` if parallel feature is disabled ([#673], [#674])
* Bump `uuid` to `0.8.1`. ([#683])
* Bump `rayon` to `1.3.0`. ([#683])

[#648]: https://github.com/amethyst/specs/pull/648
[#655]: https://github.com/amethyst/specs/pull/655
[#671]: https://github.com/slide-rs/specs/pull/671
[#673]: https://github.com/slide-rs/specs/issues/673
[#674]: https://github.com/slide-rs/specs/pull/674
[#683]: https://github.com/slide-rs/specs/pull/683

# 0.15.1 (2019-09-16)

* Benchmark uses `nalgebra` instead of `cgmath`. ([#619])
* Bumped `shrev` from `1.0` to `1.1`. ([#619]).
* Update hashbrown to 0.6.0, criterion to 0.3 ([#627], [#632])
* Remove `mopa` in favour of `std::any::Any` ([#631])

[#619]: https://github.com/slide-rs/specs/pull/619
[#627]: https://github.com/slide-rs/specs/pull/627
[#631]: https://github.com/slide-rs/specs/pull/631
[#632]: https://github.com/slide-rs/specs/pull/632

# 0.15.0 (2019-06-30)

* Removed `common` and `RudyStorage` ([#542])
* Moved `World` to `shred`, added `WorldExt` trait for Specs functionality ([#550])
* Add `UuidMarker` for UUID <-> `Entity` mappings ([#584])
* Implement `Join` on `BitSetLike` trait object ([#599])
* Expose inner field of `AntiStorage` ([#603])
* Remove `fnv` in favour of `hashbrown` ([#606])
* Reexport `hibitset`, `rayon`, `shred` and `shrev` ([#606])
* Reexport `shred_derive::SystemData` when `shred-derive` feature is enabled ([#606])
* Reexport `specs_derive::{Component, ConvertSaveload}` when `specs-derive` feature is enabled
([#606])

[#542]: https://github.com/slide-rs/specs/pull/542
[#550]: https://github.com/slide-rs/specs/pull/550
[#584]: https://github.com/slide-rs/specs/pull/584
[#599]: https://github.com/slide-rs/specs/pull/599
[#603]: https://github.com/slide-rs/specs/pull/603
[#606]: https://github.com/slide-rs/specs/pull/606

# 0.14.2 (2019-01-04)

* Add `Join`-able entries API to `Storage` ([#518])
* Several docs / meta improvements ([#526], [#527], [#528], [#530], [#531])
* Fix bug when re-killing Entity after atomic killing followed by alloc ([#533])
* Add `Storage::count` and `Storage::is_empty` ([#534])

[#518]: https://github.com/slide-rs/specs/pull/518
[#526]: https://github.com/slide-rs/specs/pull/526
[#527]: https://github.com/slide-rs/specs/pull/527
[#528]: https://github.com/slide-rs/specs/pull/528
[#530]: https://github.com/slide-rs/specs/pull/530
[#531]: https://github.com/slide-rs/specs/pull/531
[#533]: https://github.com/slide-rs/specs/pull/533
[#534]: https://github.com/slide-rs/specs/pull/534

# 0.14.1 (2018-11-23)

* Allow accessing the `EntitiesRes` fetched by a `Storage` ([#515])
* Book / API doc improvements ([#496], [#507], [#511])

[#496]: https://github.com/slide-rs/specs/pull/496
[#507]: https://github.com/slide-rs/specs/pull/507
[#511]: https://github.com/slide-rs/specs/pull/511
[#515]: https://github.com/slide-rs/specs/pull/515

# 0.14.0 (2018-10-28)

* Changed `ConvertSaveload::convert_into` and `ConvertSaveload::convert_from` ([#504])

[#504]: https://github.com/slide-rs/specs/pull/504

# 0.13.0 (2018-10-28)

This release can be skipped; please use `0.14` instead.

* Generation now internally uses the new `NonZeroI32` from `nonzero_signed`, meaning `Option<Entity>` is the same
  size as `Entity`. Note this bumps the minimum supported rust version to 1.28.0 ([#447]).
* Improved `saveload` and added custom derive for components with `Entity` ([#460])
* Removed `world::Bundle` ([#486], [#505])
* Updated Chapter 7: Setup to be more explicit, updated examples to follow that methodology ([#487])
* Added some comments to the `saveload` example ([#492])
* Updated dependency versions ([#494])

src/lib.rs
#![warn(missing_docs)]
#![deny(unsafe_op_in_unsafe_fn)]
#![deny(clippy::disallowed_types)]

//! # SPECS Parallel ECS
//!
//! This library provides an ECS variant designed for parallel execution
//! and convenient usage. It is highly flexible when it comes to actual
//! component data and the way it is stored and accessed.
//!
//! Features:
//!
//! * depending on chosen features either 0 virtual function calls or one per
//!   system
//! * parallel iteration over components
//! * parallel execution of systems
//!
//! ## High-level overview
//!
//! One could basically split this library up into two parts:
//! The data part and the execution part.
//!
//! ### The data
//!
//! `World` is where component storages, resources and entities are stored.
//! See the docs of [`World`] for more.
//!
//! [`World`]: world/struct.World.html
//!
//! [`Component`]s can be easily implemented like this:
//!
//! [`Component`]: trait.Component.html
//!
//! ```rust
//! use specs::prelude::*;
//!
//! struct MyComp;
//!
//! impl Component for MyComp {
//!     type Storage = VecStorage<Self>;
//! }
//! ```
//!
//! Or alternatively, if you enable the `"derive"` feature, you can use a
//! custom `#[derive]` macro:
//!
//! ```rust
//! # extern crate specs;
//! # extern crate specs_derive;
//! # #[cfg(not(feature = "specs-derive"))] use specs_derive::Component;
//! # #[cfg(not(feature = "specs-derive"))] use specs::prelude::*;
//!
//! # #[cfg(feature = "specs-derive")]
//! use specs::{prelude::*, Component};
//!
//! #[derive(Component)]
//! #[storage(VecStorage)] // default is `DenseVecStorage`
//! struct MyComp;
//! # fn main() {}
//! ```
//!
//! You can choose different storages according to your needs.
//!
//! These storages can be [`join`]ed together, for example joining a `Velocity`
//! and a `Position` storage means you'll only get entities which have both of
//! them. Thanks to rayon, this is even possible in parallel! See [`ParJoin`]
//! for more.
//!
//! [`join`]: trait.Join.html#method.join
//! [`ParJoin`]: trait.ParJoin.html
//!
//! ### System execution
//!
//! Here we have [`System`] and [`Dispatcher`] as our core types. Both types
//! are provided by a library called `shred`.
//!
//! [`Dispatcher`]: struct.Dispatcher.html
//! [`System`]: trait.System.html
//!
//! The `Dispatcher` can be seen as an optional part here;
//! it allows dispatching the systems in parallel, given a list
//! of systems and their dependencies on other systems.
//!
//! If you don't like it, you can also execute the systems yourself
//! by using [`RunNow`].
//!
//! [`RunNow`]: trait.RunNow.html
//!
//! `System`s are traits with a `run()` method and an associated
//! [`SystemData`], allowing type-safe aspects (knowledge about the
//! reads / writes of the systems).
//!
//! [`SystemData`]: trait.SystemData.html
//!
//! ## Examples
//!
//! This is a basic example of using Specs:
//!
//! ```rust
//! extern crate specs;
//!
//! use specs::prelude::*;
//!
//! // A component contains data which is
//! // associated with an entity.
//!
//! struct Vel(f32);
//!
//! impl Component for Vel {
//!     type Storage = VecStorage<Self>;
//! }
//!
//! struct Pos(f32);
//!
//! impl Component for Pos {
//!     type Storage = VecStorage<Self>;
//! }
//!
//! struct SysA;
//!
//! impl<'a> System<'a> for SysA {
//!     // These are the resources required for execution.
//!     // You can also define a struct and `#[derive(SystemData)]`,
//!     // see the `full` example.
//!     type SystemData = (WriteStorage<'a, Pos>, ReadStorage<'a, Vel>);
//!
//!     fn run(&mut self, (mut pos, vel): Self::SystemData) {
//!         // The `.join()` combines multiple components,
//!         // so we only access those entities which have
//!         // both of them.
//!
//!         // This joins the component storages for Position
//!         // and Velocity together; it's also possible to do this
//!         // in parallel using rayon's `ParallelIterator`s.
//!         // See `ParJoin` for more.
//!         for (pos, vel) in (&mut pos, &vel).join() {
//!             pos.0 += vel.0;
//!         }
//!     }
//! }
//!
//! fn main() {
//!     // The `World` is our
//!     // container for components
//!     // and other resources.
//!
//!     let mut world = World::new();
//!     world.register::<Pos>();
//!     world.register::<Vel>();
//!
//!     // An entity may or may not contain some component.
//!
//!     world.create_entity().with(Vel(2.0)).with(Pos(0.0)).build();
//!     world.create_entity().with(Vel(4.0)).with(Pos(1.6)).build();
//!     world.create_entity().with(Vel(1.5)).with(Pos(5.4)).build();
//!
//!     // This entity does not have `Vel`, so it won't be dispatched.
//!     world.create_entity().with(Pos(2.0)).build();
//!
//!     // This builds a dispatcher.
//!     // The third parameter of `add` specifies
//!     // logical dependencies on other systems.
//!     // Since we only have one, we don't depend on anything.
//!     // See the `full` example for dependencies.
//!     let mut dispatcher = DispatcherBuilder::new().with(SysA, "sys_a", &[]).build();
//!
//!     // This dispatches all the systems in parallel (but blocking).
//!     dispatcher.dispatch(&mut world);
//! }
//! ```
//!
//! You can also easily create new entities on the fly:
//!
//! ```
//! use specs::prelude::*;
//!
//! struct EnemySpawner;
//!
//! impl<'a> System<'a> for EnemySpawner {
//!     type SystemData = Entities<'a>;
//!
//!     fn run(&mut self, entities: Entities<'a>) {
//!         let enemy = entities.create();
//!     }
//! }
//! ```
//!
//! See the repository's examples directory for more examples.

pub extern crate hibitset;
#[cfg(feature = "parallel")]
pub extern crate rayon;
pub extern crate shred;
pub extern crate shrev;
#[cfg(feature = "uuid_entity")]
pub extern crate uuid;

#[cfg(feature = "serde")]
pub mod saveload;

mod bitset;
pub mod changeset;
pub mod error;
pub mod join;
pub mod prelude;
pub mod storage;
pub mod world;

pub use hibitset::BitSet;
pub use shred::{
    Accessor, AccessorCow, BatchAccessor, BatchController, BatchUncheckedWorld, Dispatcher,
    DispatcherBuilder, Read, ReadExpect, RunNow, RunningTime, StaticAccessor, System, SystemData,
    World, Write, WriteExpect,
};
pub use shrev::ReaderId;

#[cfg(feature = "parallel")]
pub use shred::AsyncDispatcher;

#[cfg(feature = "specs-derive")]
pub use specs_derive::{Component, ConvertSaveload};

#[cfg(feature = "parallel")]
pub use crate::join::ParJoin;
pub use crate::{
    changeset::ChangeSet,
    join::{Join, LendJoin},
    storage::{
        DefaultVecStorage, DenseVecStorage, FlaggedStorage, HashMapStorage, NullStorage,
        ReadStorage, Storage, Tracked, VecStorage, WriteStorage,
    },
    world::{Builder, Component, Entities, Entity, EntityBuilder, LazyUpdate, WorldExt},
};

pub use crate::storage::DerefFlaggedStorage;
